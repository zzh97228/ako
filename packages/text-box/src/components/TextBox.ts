import vue, { computed, defineComponent, h, reactive, VNodeArrayChildren } from 'vue';
import { genModelProps, useModel, useFieldConsumer, convertToNumber } from '@lagabu/shared';

export default defineComponent({
  name: 'TextBox',
  props: {
    autofocus: Boolean,
    minlength: [String, Number],
    maxlength: [String, Number],
    showCount: Boolean,
    disabled: Boolean,
    readonly: Boolean,
    placeholder: String,
    autocomplete: Boolean,
    spellcheck: [Boolean, String],
    rows: [String, Number],
    cols: [String, Number],
    ...genModelProps(String),
  },
  setup(props, context) {
    const state = reactive({
      isComposing: false,
    });
    const notAllowed = computed(() => props.disabled || props.readonly);

    const { model, lazyState, setInnerState } = useModel(props, context, notAllowed);
    useFieldConsumer({
      model,
      lazyState,
      setInnerState,
    });

    const onInput = (e: InputEvent) => {
      if (notAllowed.value || state.isComposing) {
        e.stopImmediatePropagation();
        return;
      }
      const target = e.target as HTMLTextAreaElement;
      model.value = target.value;
      context.emit('input', lazyState.value);
    };

    const genTextarea = () => {
      return h('textarea', {
        class: 'text-box',
        placeholder: props.placeholder,
        disabled: props.disabled,
        readonly: props.readonly,
        autofocus: props.autofocus,
        autocomplete: props.autocomplete ? 'on' : 'off',
        spellcheck: props.spellcheck,
        minlength: props.minlength,
        maxlength: props.maxlength,
        rows: props.rows || 10,
        cols: props.cols || 20,
        onInput,
        onCompositionstart: () => {
          state.isComposing = true;
        },
        onCompositionend: () => {
          state.isComposing = false;
        },
      });
    };
    return {
      model,
      lazyState,
      state,
      genTextarea,
    };
  },
  methods: {
    genCountText(count: number) {
      const children: VNodeArrayChildren = [h('span', String(count))];
      if (this.maxlength) {
        children.push(h('span', '/'));
        children.push(h('span', this.maxlength));
      }
      return h(
        'span',
        {
          class: 'text-box__count',
        },
        children
      );
    },
    genCountTextWrapper() {
      if (!this.showCount) return;
      let children: VNodeArrayChildren = [];
      const wordLength = this.lazyState.value?.length || 0;
      if (this.$slots.count) {
        children = [this.$slots.count(wordLength)];
      } else {
        children.push(this.genCountText(wordLength));
      }
      return h(
        'div',
        {
          class: 'text-box__count--wrapper',
        },
        children
      );
    },
    genBoxContent() {
      return h(
        'div',
        {
          class: 'text-box__content',
        },
        [this.genTextarea(), this.genCountTextWrapper()]
      );
    },
  },
  render() {
    return h(
      'div',
      {
        class: {
          'text-box__wrapper': true,
          'text-box--disabled': this.disabled,
        },
      },
      [this.genBoxContent()]
    );
  },
});
