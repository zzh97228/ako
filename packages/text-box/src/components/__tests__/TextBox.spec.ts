import { mount, VueWrapper } from '@vue/test-utils';
import TextBox from '../TextBox';
describe('TexyBox.ts', () => {
  let mountFunc: (opts?: object) => VueWrapper<any>;
  beforeEach(() => {
    mountFunc = (options = {}) => mount(TextBox, options);
  });

  it('should render TextBox with basic props', () => {
    const options = {
      autofocus: true,
      minlength: 12,
      maxlength: 20,
      placeholder: 'hello world',
      spellcheck: false,
      autocomplete: true,
      rows: 20,
      cols: 50,
    };
    const wrapper = mountFunc({
      props: options,
    });

    expect(wrapper.classes()).toContain('text-box__wrapper');
    const textarea = wrapper.find('textarea');
    expect(textarea.exists()).toBeTruthy();
    expect(textarea.classes()).toContain('text-box');
    expect(textarea.attributes('autofocus')).not.toBeUndefined();
    expect(textarea.attributes('minlength')).toBe(String(options.minlength));
    expect(textarea.attributes('maxlength')).toBe(String(options.maxlength));
    expect(textarea.attributes('placeholder')).toBe(options.placeholder);
    expect(textarea.attributes('spellcheck')).not.toBeUndefined();
    expect(textarea.attributes('autocomplete')).toBe('on');
    expect(textarea.attributes('rows')).toBe(String(options.rows));
    expect(textarea.attributes('cols')).toBe(String(options.cols));
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render text-box__count element when set `showCount` prop', async () => {
    const word = 'hello world';
    const wrapper = mountFunc({
      props: {
        showCount: true,
        modelValue: word,
      },
    });
    const textCount = wrapper.find('.text-box__count--wrapper');
    expect(textCount.exists()).toBeTruthy();
    expect(textCount.text()).toBe(String(word.length));
    await wrapper.setProps({
      maxlength: 20,
    });
    expect(textCount.text()).toBe(`${word.length}/${20}`);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should not set value when `disabled` prop set', async () => {
    const wrapper = mountFunc({
      props: {
        modelValue: 'hello world',
        disabled: true,
      },
    });
    expect(wrapper.vm.lazyState.value).toBe('hello world');
    wrapper.vm.model = 'hello there';
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.lazyState.value).toBe('hello world');
    await wrapper.setProps({
      modelValue: 'disabled',
    });
    expect(wrapper.vm.lazyState.value).toBe('hello world');
  });
});
