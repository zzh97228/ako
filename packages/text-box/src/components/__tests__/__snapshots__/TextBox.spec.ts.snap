// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TexyBox.ts should render TextBox with basic props 1`] = `
<div class="text-box__wrapper">
  <div class="text-box__content">
    <textarea class="text-box"
              placeholder="hello world"
              autofocus
              autocomplete="on"
              spellcheck="false"
              minlength="12"
              maxlength="20"
              rows="20"
              cols="50"
    >
    </textarea>
  </div>
</div>
`;

exports[`TexyBox.ts should render text-box__count element when set \`showCount\` prop 1`] = `
<div class="text-box__wrapper">
  <div class="text-box__content">
    <textarea class="text-box"
              autocomplete="off"
              spellcheck="false"
              rows="10"
              cols="20"
              maxlength="20"
    >
    </textarea>
    <div class="text-box__count--wrapper">
      <span class="text-box__count">
        <span>
          11
        </span>
        <span>
          /
        </span>
        <span>
          20
        </span>
      </span>
    </div>
  </div>
</div>
`;
